aptos-move/block-executor/src/view.rs

```diff
@@ -811,14 +1159,123 @@ impl<'a, T: Transaction, S: TStateView<Key = T::Key>, X: Executable> LatestView<
+                    return None;
+                }
+
+                // TODO[agg_v2](fix): Is it guaranteed that metadata is not None?
+                if let Ok(Some(maybe_metadata)) = self.get_resource_state_value_metadata(&key) {
+                    let maybe_metadata = maybe_metadata.map(
+                        |metadata: aptos_types::state_store::state_value::StateValueMetadata| {
+                            StateValue::new_with_metadata(Bytes::new(), metadata)
+                        },
                     );
+                    if let Ok(GroupReadResult::Size(group_size)) =
+                        parallel_state.read_group_size(&key, self.txn_idx)
+                    {
+                        let metadata_op: T::Value =
+                            TransactionWrite::from_state_value(maybe_metadata);
+                        if let Some(metadata_op) = metadata_op.convert_read_to_modification() {
+                            return Some(Ok((key.clone(), (metadata_op, group_size))));
+                        }
+                    }
+                }
+                // TODO[agg_v2](fix): Is this a code invariant error?
+                Some(Err(code_invariant_error(format!(
+                    "Cannot metdata op for the group read {:?}",
+                    key
+                ))))
+            })
+            .collect()
+    }
+
+    fn get_group_reads_needing_exchange_sequential(
+        &self,
+        group_read_set: &HashSet<T::Key>,
+        unsync_map: &UnsyncMap<T::Key, T::Tag, T::Value, X, T::Identifier>,
+        delayed_write_set_keys: &HashSet<T::Identifier>,
+        skip: &HashSet<T::Key>,
+    ) -> Result<BTreeMap<T::Key, (T::Value, u64)>, PanicError> {
+        group_read_set
+            .iter()
+            .filter(|key| !skip.contains(key))
+            .flat_map(|key| {
+                if let Some(value_vec) = unsync_map.fetch_group_data(key) {
+                    let mut resources_needing_delayed_field_exchange = false;
+                    for (_, value_with_layout) in value_vec {
+                        if let ValueWithLayout::Exchanged(value, Some(layout)) = value_with_layout {
+                            if let Some(bytes) = value.bytes() {
+                                let identifiers_in_read = self
+                                    .extract_identifiers_from_value(bytes, layout.as_ref())
+                                    .unwrap();
+                                if !delayed_write_set_keys.is_disjoint(&identifiers_in_read) {
+                                    resources_needing_delayed_field_exchange = true;
+                                    break;
+                                }
+                            }
+                        }
+                    }
+                    if !resources_needing_delayed_field_exchange {
+                        return None;
+                    }
+                    if let Some(metadata) = unsync_map.fetch_data(key) {
```

Comment:
this isn't metadata? isn't it actual value?

Reply:
this is metadata of the group - it is what is stored in regular value for the resource group state key?

Reply:
Good point, nevermind then. The minor confusion for me was that there is also metadata type (which is in fact contained in the value here), but I can't necessarily think of a much better name